sql/core/src/main/scala/org/apache/spark/sql/execution/offset.scala

```diff
@@ -0,0 +1,66 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.serializer.Serializer
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.{Attribute, SortOrder}
+import org.apache.spark.sql.catalyst.plans.physical.{Partitioning, SinglePartition}
+
+
+/**
+ * Skip the first `offset` elements and collect them to a single partition.
+ * This operator will be used when a logical `Offset` operation is the final operator in an
+ * logical plan, which happens when the user is collecting results back to the driver.
+ */
+case class CollectOffsetExec(offset: Int, child: SparkPlan) extends UnaryExecNode {
+
+  override def output: Seq[Attribute] = child.output
+
+  override def outputPartitioning: Partitioning = SinglePartition
+
+  override def outputOrdering: Seq[SortOrder] = child.outputOrdering
+
+  override def executeCollect(): Array[InternalRow] = child.executeCollect.drop(offset)
+
+  private val serializer: Serializer = new UnsafeRowSerializer(child.output.size)
+
+  protected override def doExecute(): RDD[InternalRow] = {
+    sparkContext.parallelize(executeCollect(), 1)
+  }
+
+}
+
+/**
+ * Skip the first `offset` elements and collect them to a single partition.
+ */
+case class OffsetExec(offset: Int, child: SparkPlan) extends UnaryExecNode {
+
+  override def output: Seq[Attribute] = child.output
+
+  override def outputOrdering: Seq[SortOrder] = child.outputOrdering
+
+  protected override def doExecute(): RDD[InternalRow] = {
+    val rdd = child.execute()
+    val arr = rdd.take(offset)
+    rdd.filter(!arr.contains(_))
```

Comment:
We need some design about how to implement OFFSET on a distributed data set. The current approach works, but it's not robust as it needs to broadcast `arr` (which can be large if OFFSET is large) and relies on object equality.

We may need to follow how LIMIT is implemented in Spark.

Reply:
Yes, I think so.
I have referenced the implement of `LIMIT`, but `OFFSET` looks can't follow the same way as `LIMIT`.
Second, `OFFSET` easier to generate large amounts of data than `LIMIT`.
I have an immature suggestion give a limitation on 'OFFSET '.

Reply:
A rough idea:
1. get the numRecords of the first partition
2. If the numRecords is bigger than OFFSET, go to step 4
3. get numRecords of more partitions (quadruple and retry like LIMIT), until total numRecords is bigger than OFFSET.
4. Now we have the numRecords of some head partitions that totoal numRecords exceeds the OFFSET, we can easily skip the head records.

If we have accurate per-partition numRecords statistics, we can go step 4 directly.

Reply:
@cloud-fan It is a better idea for me. I will reactor the current implement.

Reply:
@cloud-fan There are exists a problem the index of partition and the order of data are inconsistent.
I have a new implement but not works file as I can't assurance the order of output produced by child plans.

Reply:
So, I want know how to assurance the order?